Backpack/Chip/Classes/BPKChip.m

```diff
@@ -0,0 +1,161 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "BPKChip.h"
+
+#import <Backpack/Font.h>
+#import <Backpack/Label.h>
+#import <Backpack/Color.h>
+#import <Backpack/Shadow.h>
+#import <Backpack/Spacing.h>
+#import <Backpack/UIView+BPKRTL.h>
+
+NS_ASSUME_NONNULL_BEGIN
+@interface BPKChip()
+@property(nonatomic, getter=isInitializing) BOOL initializing;
+@property(nonatomic) BPKLabel* titleLabel;
+
+@property(nonatomic) UIColor *textColor;
+@end
+
+@implementation BPKChip
+
+- (instancetype)initWithFrame:(CGRect)frame {
+    self = [super initWithFrame:frame];
+    
+    if (self) {
+        [self setup];
+    }
+    
+    return self;
+}
+
+- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder {
+    self = [super initWithCoder:aDecoder];
+    
+    if (self) {
+        [self setup];
+    }
+    
+    return self;
+}
+
+- (void)setup {
+    self.initializing = YES;
+    
+    self.titleLabel = [[BPKLabel alloc]initWithFrame:CGRectZero];
+    [self.titleLabel setFontStyle:BPKFontStyleTextSm];
+    [self addSubview:self.titleLabel];
+    
+    self.titleLabel.textAlignment = NSTextAlignmentCenter;
+    
+    UITapGestureRecognizer *singleFingerTap =
+    [[UITapGestureRecognizer alloc] initWithTarget:self
+                                            action:@selector(handleSingleTap:)];
+    [self addGestureRecognizer:singleFingerTap];
+    
+    [self updateStyle];
+    self.initializing = NO;
+}
+
+#pragma mark - Style setters
+
+- (void)setTitle:(NSString *_Nullable)title {
+    _title = [title copy];
+    
+    [self updateTitle];
+}
+
+#pragma mark - State setters
+
+- (void)setSelected:(BOOL)selected {
+    [super setSelected:selected];
+    [self updateStyle];
+}
+
+- (void)setEnabled:(BOOL)enabled {
+    [super setEnabled:enabled];
+    [self updateStyle];
+}
+
+- (void)setHighlighted:(BOOL)highlighted {
+    [super setHighlighted:highlighted];
+}
+
+#pragma mark - Layout
+
+- (void)layoutSubviews {
+    [super layoutSubviews];
+    
+    [self setConstraints];
```

Comment:
This will re create all constraints everytime the view is laid out. This is pretty bad for performance. Can we create the constriant in `setup` instead?